### ISSUE TITLE
Edge fitting

### ISSUE THREAD

**Issue by ZhouCX117** [kunalchelani/EdgeGaussians#1]
Thanks for your excellent! How could I use the EMAP code to fit curves and lines from the trained 3DGS? Now I regard the Gaussian centers and directions as points and linedirection and feed them into the get_parametric_edge function of EMAP [https://github.com/cvg/EMAP/blob/main/src/edge_extraction/extract_parametric_edge.py](url). Is it right?

**Comment by kunalchelani** [comment]
Hello! Yes, that would be one way to do it. However, EMAP's method of clustering of points with direction, and also ours is fitted to the distribution of points and associated direction (and noise in it) obtained from the underlying process (voxel-based UDF and Gaussians respectively). So, if you haven't applied the geometric regularization to the Gaussians mentioned in our paper, you would get a slightly different distribution of Gaussian centers. 
It is however worth trying out the steps you mention to establish one baseline. Curious to know how it goes!

**Comment by ZhouCX117** [comment]
I have applied the geometric regularization to the initial Gaussians and utilized the EMAP's code to fit curves and lines. The results are bad. It seems the trained Gaussians are noisy. Could you please provide the code described in Section 3.3?

**Comment by kunalchelani** [comment]
We will be releasing the code soon. Thanks for the patience!

**Comment by Jiahao-Ma** [comment]
Thanks for your awesome work. I am still curious about edge filtering. Some Gaussians do not lie on the edge. Should we first filter and select only those Gaussians on the edge before applying EMAP's method to extract the edge, or should we apply EMAP's method to all Gaussians?

**Comment by kunalchelani** [comment]
Filtering is indeed helpful if there are wayward Gaussians. You can also fit edges to all Gaussians and then filter out edges that do not project to edges in the 2D edge maps, how it was done in EMAP. In our work, we performed opacity based filtering, along with filtering statistical outliers. For ABC-NEF objects, we also filtered Gaussians not projecting to an edge in at least a minimum percentage (5-10%) of total views.

**Comment by kunalchelani** [comment]
Added code. Please feel free to re-open if there are more issues, or you'd want to discuss.

**Comment by ZhouCX117** [comment]
@kunalchelani Thanks for your sharing. I am trying to reproduce the results. However, in line 36 of 'fit_edge.py', the function 'init_views_and_get_scale' is not available in the train_utils.py file. Could you please check it? Thanks again!

**Comment by kunalchelani** [comment]
Thanks for pointing out. I had missed a refactoring change. Let me know if there are still issues with the latest commit.

**Comment by ZhouCX117** [comment]
@kunalchelani Thanks for your reply. It works now! Excellent results! By the way, could you please give a more detailed guidance for evaluation? Step1: Edge extraction and Step 2: eval.py? How about the config setup such as 'use_parametric_edges' and 'use_filtered_points'? And is the process same for ABC and DTU datasets?

**Comment by ZhouCX117** [comment]
Sorry to bother you again. Could you please provide the JSON file for both PidiNet and DexiNed edge detectors?

**Comment by kunalchelani** [comment]
Regarding the evaluation, you should use the flag --use_parametric_edges as it uses the points sampled at the desired resolution on the parametric edges. The flag --use_filtered_points is for evaluating the raw Gaussian means, but they can be denser or sparser than the ground truth sampling and therefore this evaluation is not completely correct to do. I will remove this option in the next commit.
The eval.py in its current state is meant for evaluation over ABC. As mentioned in the README, evaluation over DTU can done using the referred EMAP script.

As far as the configs are concerned, I will upload the PidiNet config for ABC after testing that it produces roughly the same results as mentioned in the paper, but I can do this some time next week.

**Comment by ZhouCX117** [comment]
@kunalchelani Thanks for your reply. Looking forward to seeing the update. By the way, could you please provide the render file?

**Comment by ZhouCX117** [comment]
@kunalchelani Thank you for your response! I need a file similar to https://github.com/graphdeco-inria/gaussian-splatting/blob/main/render.py for visualization.

**Comment by kunalchelani** [comment]
I see, I currently don't save the rendered images from the views. I might add the functionality later. I guess you can use something similar to [get_output function](https://github.com/kunalchelani/EdgeGaussians/blob/e35c641f0ed0056e600c4c18d158235af83bbe1d/edgegaussians/models/edge_gs.py#L197) after [loading the viewpoints ](https://github.com/kunalchelani/EdgeGaussians/blob/e35c641f0ed0056e600c4c18d158235af83bbe1d/edgegaussians/utils/data_utils.py#L11)and [learnt Gaussian params](https://github.com/kunalchelani/EdgeGaussians/blob/e35c641f0ed0056e600c4c18d158235af83bbe1d/edgegaussians/models/edge_gs.py#L625), to render the images.

**Comment by kunalchelani** [comment]
@Chauncywu1999 The warning is most likely not the cause of not getting curves. Are you sure you are using the correct config file for DTU (config/DTU.json) ? The parameters `line_curve_residual_comp_factor` and `angle_thresh` in the parametric_fitting section of the config affect the inclination towards more lines or edges. Upon testing at my end, I get reasonable number of curves with the default configuration in DTU.json.

**Comment by Chauncywu1999** [comment]
@kunalchelani Yes, I did not change the DTU.json. This is the result of edge extraction that I have achieved so far, which feels like there are fewer lines compared to the original paper.
![屏幕截图 2024-11-22 091517](https://github.com/user-attachments/assets/25248e89-ce5f-48f9-a512-427853a754a1)
<img src="https://github.com/user-attachments/assets/9d730d23-b489-4f1e-88cd-9e584d332c6c" alt="Example Image" width="300">

**Comment by kunalchelani** [comment]
This looks quite similar to the output in the paper to me. Perhaps, a few edges less which can be because of the duplication and culling, which at the moment is at fixed steps. This is one of the shortcomings that the current pipeline is prone to hyper-parameter tuning but that is the case for the original Gaussian Splatting paper as well, on which this is based. Anyway, you can try with this config, gives a few more edges (I haven't checked the reproducibility for DTU yet, will do that soon). [DTU.json](https://github.com/user-attachments/files/17868660/DTU.json)

**Comment by ZhouCX117** [comment]
@kunalchelani Hi, would you please provide the trained checkpoints and the JSON files for the TT and Relipca datasets?

**Comment by ZhouCX117** [comment]
@kunalchelani Sorry to bother you again. I want to replace the rasterization method from gsplat to the initial Gaussian Splatting, but I met some errors. For the grad update, I replace the self.xys.absgrad with viewspace_point_tensor.grad. After some steps, the points become nan. Could you please give me a favor? Thanks again for your help!

**Comment by kunalchelani** [comment]
@ZhouCX117 The results for replica can be obtained using the provided config file. I will soon add the Tnt results as well.
Regarding replacing the gsplat based rasterization to the original one, I don't have much idea as I haven't used the original one myself. I hope you're able to solve it though. Closing this discussion, please feel free to open a new issue in case something else pops up.
